File: JavaAlgorithm/src/main/java/algo/tzashinorpu/SpecialSubject/Top/myAtoi_8.java
package algo.tzashinorpu.SpecialSubject.Top;

import java.util.HashMap;
import java.util.Map;

public class myAtoi_8 {
   /* DFA
    	        ' '	    +/-	    number	   other
    start	    start	signed	in_number	end
    signed	    end	    end	    in_number	end
    in_number	end	    end	    in_number	end
    end	        end	    end	    end	        end*/

    public int myAtoi(String s) {
        Automaton automaton = new Automaton();
        for (int i = 0; i < s.length(); i++) {
            automaton.get(s.charAt(i));
        }
        return (int) (automaton.sign * automaton.ans);
    }

    class Automaton {
        public int sign = 1;
        public long ans = 0;
        private String state = "start";
        private Map<String, String[]> table = new HashMap<>() {{
            put("start", new String[]{"start", "signed", "in_number", "end"});
            put("signed", new String[]{"end", "end", "in_number", "end"});
            put("in_number", new String[]{"end", "end", "in_number", "end"});
            put("end", new String[]{"end", "end", "end", "end"});
        }};

        public void get(char c) {
            state = table.get(state)[getCol(c)];
            if (state.equals("in_number")) {
                ans = ans * 10 + c - '0';
                ans = sign == 1 ? Math.min(ans, (long) Integer.MAX_VALUE) : Math.min(ans, -(long) Integer.MIN_VALUE);
            } else if (state.equals("signed")) {
                sign = c == '+' ? 1 : -1;
            }
        }

        private int getCol(char c) {
            if (c == ' ') return 0;
            if (c == '+' || c == '-') return 1;
            if (Character.isDigit(c)) return 2;
            return 3;
        }
    }
}

File: Cpp/Algorithm/src/SpecialSubject/MyAtoi_8.cpp
#include "ss.h"

class Automaton
{
public:
	Automaton();
	~Automaton();
	int sign = 1;
	long ans = 0;
	void get(char c) {
		state = table[state][getCol(c)];
		if (state == "number")
		{
			ans = ans * 10 + c - '0';
			ans = sign == 1 ? min(ans, (long)INT_MAX) : min(ans, -1 * (long)INT_MIN);
		}
		if (state == "signed" && c == '-')
		{
			sign = -1;
		}
	}

private:
	string state = "start";
	unordered_map<string, vector<string>> table = {
		{"start", {"start", "signed", "number", "end"}},
		{"signed", {"end", "end", "number", "end"}},
		{"number", {"end", "end", "number", "end"}},
		{"end", {"end", "end", "end", "end"}}
	};
	int getCol(char c)
	{
		if (isspace(c)) return 0;
		if (c == '+' || c == '-') return 1;
		if (isdigit(c)) return 2;
		return 3;
	}

};

Automaton::Automaton()
{
}

Automaton::~Automaton()
{
}
class MyAtoi_8 {
private:
	/*
				" "				"+"/"-"       number        other
	started		started			signed		  number		end
	signed		end				 end		  number		end
	number		end				 end          number		end
	end			end				 end          end			end
	*/
public:
	int myAtoi(string s) {
		Automaton automaton;
		for (char c : s)
		{
			automaton.get(c);
		}
		return automaton.sign * automaton.ans;
	}
};
